### ISSUE TITLE
getMask for 1.15.2 returns wrong value in some cases

### ISSUE THREAD

**Issue by kf106** [PrismarineJS/prismarine-chunk#205]
In 1.15.2 if you build up a chunk, then delete the blocks (replacing them with air), the chunk.getMask value returned is that for the previous tall chunk, i.e. it doesn't recompute the mask to only count upward to the last non-air block.

This means that if you build a tall tower, delete it with mining, save the chunk using chunk.dump, save the bitmask value using chunk.getMask, and then try to load the saved chunk using chunk.load with the stored getMask value, you get an error that you're reading past the memory bound.

**Comment by nickelpro** [comment]
So two things:
* We never re-compress a chunk (which is something we can argue about the correctness of)
* Therefore, once a block has been set inside a chunk, _that chunk is always part of the `sectionMask`_. Once a chunk is non-empty, it is forever non-empty from the point of view of pris-chunk. We never zero any bits in the `sectionMask`

So `.getMask()` at runtime is correct, no bug there, because internally that's how the chunk is calculating. Ie, a section object with a palette exists for that part of the chunk column.

The problem is how we serialize in `.dump()`, we ask the wrong question, we ask:
```js
if (section !== null && !section.isEmpty()) {
  section.write(smartBuffer)
}
```

Ostensibly as a space-saving measure. And _this_ is wrong, or at least a tricky corner. I think we need to rethink this. It might be we need do a naive compression pass (eliminating empty chunks and recalculating the section mask) whenever `dump()` is called, but we need some way to communicate that `getMask()` calls from prior to the `dump()` are invalid after the `dump()`.

**Comment by nickelpro** [comment]
I'm not looking at any docs here, is it invalid to serialize a chunk with zero solid blocks? I don't think so.

The bug fix might be that we remove the `!section.isEmpty()` check from affected versions. And add a separate `compress()` method to their chunk column objects that does this removal of empty sections and recalcs the section mask.

Ping @extremeheat

**Comment by kf106** [comment]
I tried implementing your suggestion, and indeed there are other areas which subsequently fail if you update the section mask.

So I took the quick and easy route and added a new function to get the actual mask for the dumped buffer:

```
    dumpMask () {
      let newMask = this.sectionMask
      this.sections.forEach((section, i) => {
        if (section !== null && section.isEmpty()) {
          newMask &= ~(1 << i)
        }
      })
      return newMask
    }

```

When I'm using dump and load in my fs plugin, instead of calling getMask I call dumpMask and get the correct value, and the rest of the code works as before.

**Comment by extremeheat** [comment]
> Ostensibly as a space-saving measure. And this is wrong, or at least a tricky corner. I think we need to rethink this. It might be we need do a naive compression pass (eliminating empty chunks and recalculating the section mask) whenever dump() is called, but we need some way to communicate that getMask() calls from prior to the dump() are invalid after the dump().

Yeah, on the bedrock implementation we have some handling for this:

https://github.com/PrismarineJS/prismarine-chunk/blob/23ae199ad600791a67bed264cbf6af496fed7520/src/bedrock/1.3/SubChunk.js#L282-L297

And compact() on serializatized export:

https://github.com/PrismarineJS/prismarine-chunk/blob/23ae199ad600791a67bed264cbf6af496fed7520/src/bedrock/1.3/SubChunk.js#L166

Works by tracking the count of the items in the palette and rebuilding the storage at the end (if needed) when something like dump() would be called.
